api/src/test/java/io/opentelemetry/trace/SpanBuilderTest.java

```diff
@@ -32,8 +33,8 @@
   void doNotCrash_NoopImplementation() {
     Span.Builder spanBuilder = tracer.spanBuilder("MySpanName");
     spanBuilder.setSpanKind(Kind.SERVER);
-    spanBuilder.setParent(DefaultSpan.getInvalid());
-    spanBuilder.setParent(DefaultSpan.getInvalid().getContext());
+    spanBuilder.setParent(TracingContextUtils.withSpan(DefaultSpan.create(null), Context.ROOT));
```

Comment:
youch. this API sure is ugly! Remind me again how this is going to make things easier/better for our users?

Reply:
This is a unit test. Normally you would just call setNoParent.

Reply:
fair enough, but I think having to do parenting via the TCU is going to be pretty counter-intuitive to most users. :(

Reply:
I fully agree, but at the same time, I don't think users should normally do that. Usually you would just use the implicit current parent, or sometimes you need to capture a context with `Context.current()` and forward that to somewhere else. If you start a span at one thread and need to continue it at another thread while *not* setting it as active in the current thread is the only place you would need TracingContextUtils. I also think there should be some may to make that more easy. I could imagine `span.inCurrentContext()` (Java 8 default methods?), or `TracingContextUtils.spanInCurrent(span)` to make this easier.

> Remind me again how this is going to make things easier/better for our users?

As you said, using TracingContextUtils is rather unintuitive, and thus I think most users did not do it, they just passed around spans without the remaining context. However, a custom propagator should be able to set a value in the context in extract and rely on it still being available in inject.

BTW, using `inject` is a place where you already have to often muck around with TracingContextUtils currently.

Reply:
Is it safe to assume the implicit current parent is accurate?

For instance, when dealing with asynchronous methods the thread being executed on is usually not the same thread that might have initiated/received the original request. In that situation, the implicit current parent from `ThreadLocal` is more than likely not the parent you want

Reply:
No, in that case you need to capture a Context (ideally you can use Context.current()) at he parent side. There is no fundamental change here with this PR.

Reply:
If the APIs are geared towards users utilizing "implicit" parents, is it defined anywhere in what situations it results in problems and alternatives should be implemented, such as you describe?

Reply:
> If the APIs are geared towards users utilizing "implicit" parents

Hmm, I'm not sure yet why you think the API is more geared towards implicit parent's than before? I mean, I agree that it *is* a bit more cumbersome to have a span that was not active be the parent span at another thread, but the difference is only this:

```java
Span capturedParent = span;
```

vs.

```
Context capturedParent = TracingContextUtils.withSpan(span, Context.current());
```

In case the parent is active on the original thread, you can even do

```
Context capturedParent = Context.current();
```

>  is it defined anywhere in what situations it results in problems

It currently uses grpc Context, so unless some other component does some automatic in-process context propagation, it is just a thread local. I guess you should only use an explicit parent if (a) you extract one from a remote process or (b) your parent comes from a remote thread, i.e. this is the first span that continues some asynchronous operation in another "execution context" (think thread) than the logical parent span.

Reply:
Maybe we can provide some convenience function (default interface implementation?), so one can do ` Context capturedParent = span.withCurrentContext()`.

Reply:
Personally, it's a lot messier, but more importantly way less obvious that that's what is needed to achieve the desired result.

With respect to a default interface implementation, that wouldn't be possible with the JDK 7 requirement, correct?

Reply:
(FWIW, we need more convenience functions in `TracingContextUtils`, or in whatever new name it ends up having ;) )

Reply:
@kenfinnigan Java 7 was dropped recently, Java 8 is now required.

Reply:
Thanks @Oberon00, don't follow things for a few weeks, and everything changes!